### PR TITLE
Refactor WC_Payment_Gateway_WCPay Part 2

### PR DESCRIPTION
**Pull request by leonardola**
#### Changes proposed in this Pull Request
Related to #823 
This PR is the second part of a series of refactors of the WC_Payment_Gateway_WCPay class.

#### Testing instructions
- As a shopper add a product to the cart
- Make sure you can see the card inputs
- Buy the product
- Test again using UPE
- As a shopper buy any product using WooPay

- Change and save WC Pay settings via wp-admin and make sure they are saved correctly and reflect in the shopper flow

- Execute other [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) that you think are necessary

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/class-wc-payments-gateway-wcpay-settings.php
Comment (RadoslavGeorgiev) [inline includes/class-wc-payments-gateway-wcpay-settings.php]: Would you consider using namespaces here and simply calling the class `Settings` within the `WCPay` namespace?

File: includes/class-wc-payments-gateway-wcpay-settings.php
Comment (RadoslavGeorgiev) [inline includes/class-wc-payments-gateway-wcpay-settings.php]: Let's name this settings, as options are still something slightly different.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (RadoslavGeorgiev) [inline includes/class-wc-payment-gateway-wcpay.php]: Do you think a more drastic approach where **all** options are accessed through `WC_Payments_Gateway_WCPay_Settings` might make sense? Here's what I mean:

1. `WC_Payments_Gateway_WCPay_Settings` is constructed with a dependency on the gateway in order to be able to access `$this->gateway->get_option()` for settings/as a fallback.
2. All usages of `WC_Payments()->get_gateway()->get_option()` are replaced with proper getters within the settings class, ex. `WC_Payments()->get_settings()->is_enabled()`.
3. This method (`get_option` here) is completely removed.

I know this will involve a lot of file changes (ex. there are 36 changed files in https://github.com/Automattic/woocommerce-payments/pull/4986 just for `is_in_test_mode()` and `is_in_dev_mode()`), but I see a few benefits:

- ~It would not matter~ The gateway would not need to know where the option comes from (cached account data vs gateway settings).
- All options will be clearly accessible from a single place.
- We'd be able to utilize proper types/casting.

Expanding a bit further, we could also use the field types from `WC_Payments_Gateway_WCPay_Settings::get_form_fields()` and parse the value, ending the annoying `'yes' === $gateway->get_option( 'something' )`, which also premises bugs (imagine assuming it's a boolean, then even 'no' will be true). This can be done either through magic methods (`__get` for `$settings->enabled`) and PHPDoc `@property` hints (see [this example](https://github.com/Automattic/woocommerce-payments/pull/4986/files#diff-1f0e10aa8fe69406fae8d3d26368dd8b46e67c541485b5a84669086bb0150f2fR16)), or explicit methods like this:

```php
/**
 * @property-read bool $enabled Indicates if the gateway is enabled.
 */
class {
	private function get_option( $name ) {
		$value = $this->gateway->get_option( $name );

		// Of course, this formatting would change.
		$type = $this->get_form_fields()[ $name ]['type'];
		switch ( $type ) {
			case 'checkbox':
				$value = ( 'yes' === $value );
				break;
		}

		return $value;
	}

	public function is_enabled() : bool {
		return $this->get_option( 'enabled' );
	}

	public function __get( $name ) {
		if ( isset( $this->get_form_fields()[ $name ] ) ) {
			return $this->get_option( $name );
		} else {
			// throw exception
		}
	}
}
```

This example combines both approaches, but we should stick to only one of them.

**Review by RadoslavGeorgiev — COMMENTED** [review]
Nice work! I have a couple of suggestions :)

File: includes/class-wc-payment-gateway-wcpay.php
Comment (brianyu28) [inline includes/class-wc-payment-gateway-wcpay.php]: As a naming convention, I believe we generally use either `wcpay` (no underscore) or `wc_payments` (with underscore). For consistency, what do you think about `wcpay_settings` here instead of `wc_pay_settings`?

File: includes/class-wc-payments-order-service.php
Comment (brianyu28) [inline includes/class-wc-payments-order-service.php:840]: Given that we need test mode information every time this function is called, I'd suggest making it its own parameter for this function instead of including it in `options`.

**Review by brianyu28 — COMMENTED** [review]
Thanks @leonardola, overall this looks good to me! A couple of small comments in addition to what's been mentioned in the previous review. Though per paJDYF-6lv-p2, should we consider waiting to merge these changes?

File: includes/class-wc-payments-gateway-wcpay-settings.php
Comment (leonardola) [reply includes/class-wc-payments-gateway-wcpay-settings.php]: Fixed by e3a64238dcd23f440e9fed479e94c18d3d53783c

File: includes/class-wc-payments-gateway-wcpay-settings.php
Comment (leonardola) [reply includes/class-wc-payments-gateway-wcpay-settings.php]: Fixed by 58e0a75b6c0f63f485ade4841c4fb4d40185016a

File: includes/class-wc-payment-gateway-wcpay.php
Comment (leonardola) [reply includes/class-wc-payment-gateway-wcpay.php]: `WC_Payment_Gateway_WCPay` already depend on `Settings` if we make `Settings` depend on `WC_Payment_Gateway_WCPay` then we would have a circular reference. 

This could be avoided by using `WC_Payments::get_gateway_settings` but that's not the most elegant way to do this. Though we already have code like that in other places. So I am not totally against it but not a big fan either. 

What do you think?

**Comment by leonardola** [comment]
@brianyu28 
> Though per paJDYF-6lv-p2, should we consider waiting to merge these changes?

For sure. I'll probably only merge this next year after things get back to normal. This is not a urgent change

File: includes/class-wc-payment-gateway-wcpay.php
Comment (leonardola) [reply includes/class-wc-payment-gateway-wcpay.php:147]: Fixed by 3f55c6ceb8bc74a66093e63f286b9967c025a36b

File: includes/class-wc-payments-order-service.php
Comment (leonardola) [reply includes/class-wc-payments-order-service.php:840]: Fixed by 650d2cbb61b6507909a4d643ddc54d2b3ae21a61

File: includes/class-wc-payment-gateway-wcpay.php
Comment (leonardola) [reply includes/class-wc-payment-gateway-wcpay.php]: Fixed by 3f55c6ceb8bc74a66093e63f286b9967c025a36b

**Comment by leonardola** [comment]
I added a blocked label so we don't merge this before January

**Comment by leonardola** [comment]
@alefesouza this PR is now unblocked and is ready to be reviewed

**Review by alefesouza — APPROVED** [review]
LGTM :shipit:

**Comment by leonardola** [comment]
Blocked by pdpOw2-Pn-p2 till version 5.5 is launched

**Pull request closed without merging**
